target/linux/rtd1xxx/files-5.15/drivers/clk/realtek/clk-pll-mmc.c

```diff
@@ -0,0 +1,308 @@
+{
+	clk_regmap_update(&clkm->clkr, clkm->pll_ofs, 0x00000002, val << 1);
+}
+
+static inline uint32_t get_phsel(struct clk_pll_mmc *clkm, int id)
+{
+	uint32_t sft = id ? 8 : 3;
+
+	return (clk_regmap_read(&clkm->clkr, clkm->pll_ofs) >> sft) & 0x1f;
+}
+
+static inline void set_phsel(struct clk_pll_mmc *clkm, int id, uint32_t val)
+{
+	uint32_t sft = id ? 8 : 3;
+	//printk(KERN_ERR "%s: sft=%d, id=%d, val=0x%x\n", __func__, sft, id, val);
+	clk_regmap_update(&clkm->clkr, clkm->pll_ofs, 0x1f << sft, val << sft);
+}
+
+static inline void set_sscpll_rs(struct clk_pll_mmc *clkm, uint32_t val)
+{
+	clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 4, 0x00001c00, val << 10);
+}
+
+static inline void set_sscpll_icp(struct clk_pll_mmc *clkm, uint32_t val)
+{
+	clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 4, 0x000003e0, val << 5);
+}
+
+static inline void set_ssc_div_ext_f(struct clk_pll_mmc *clkm, uint32_t val)
+{
+        clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 4, 0x03ffe000, val << 13);
+}
+static inline void set_pi_ibselh(struct clk_pll_mmc *clkm, uint32_t val)
+{
+        clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 4, 0x18000000, val << 27);
+}
+static inline void set_ssc_div_n(struct clk_pll_mmc *clkm, uint32_t val)
+{
+	clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 8, 0x03ff0000, val << 16);
+}
+static inline void set_ssc_div_n_2(struct clk_pll_mmc *clkm, uint32_t val)
+{
+        clk_regmap_update(&clkm->clkr, clkm->pll_ofs + 8, 0x00ff0000, val << 16);
+}
+static inline uint32_t get_ssc_div_n(struct clk_pll_mmc *clkm)
+{
+	return (clk_regmap_read(&clkm->clkr, clkm->pll_ofs + 8) >> 16) & 0x3ff;
+}
+
+static inline void set_pow_ctl(struct clk_pll_mmc *clkm, uint32_t val)
+{
+	clk_regmap_write(&clkm->clkr, clkm->pll_ofs + 0xc, val);
+}
+
+static inline uint32_t get_pow_ctl(struct clk_pll_mmc *clkm)
+{
+	return clk_regmap_read(&clkm->clkr, clkm->pll_ofs + 0xc);
+}
+
+static int clk_pll_mmc_phase_set_phase(struct clk_hw *hw, int degrees)
+{
+	struct clk_hw *hwp = clk_hw_get_parent(hw);
+	struct clk_pll_mmc *clkm = to_clk_pll_mmc(hwp);
+	int phase_id = (hw - &clkm->phase0_hw) ? 1 : 0;
+	uint32_t val;
+
+	val = DIV_ROUND_CLOSEST(degrees * 100, 1125);
+	set_phsel(clkm, phase_id, val);
+
+	//pr_err("%pC: %s: id=%d, degrees=%d, reg=%08x\n", hw->clk, __func__, phase_id, degrees, val);
+	udelay(10);
+	return 0;
+}
+
+static int clk_pll_mmc_phase_get_phase(struct clk_hw *hw)
+{
+	struct clk_hw *hwp = clk_hw_get_parent(hw);
+	struct clk_pll_mmc *clkm = to_clk_pll_mmc(hwp);
+	int phase_id = (hw - &clkm->phase0_hw) ? 1 : 0;
+	uint32_t val;
+
+	val = get_phsel(clkm, phase_id);
+	val = DIV_ROUND_CLOSEST(val * 360, 32);
+
+	pr_debug("%pC: %s: id=%d, degrees=%d\n", hw->clk, __func__, phase_id, val);
+	return val;
+}
+
+const struct clk_ops clk_pll_mmc_phase_ops = {
+	.set_phase = clk_pll_mmc_phase_set_phase,
+	.get_phase = clk_pll_mmc_phase_get_phase,
+};
+EXPORT_SYMBOL_GPL(clk_pll_mmc_phase_ops);
```

Comment:
would be used by eMMC driver to tune phase in order to get better performance.

any better idea to avoid exporting the symbol but can be used by other driver ?